Negative half-cell models

DrSOKane opened the pull request:
# Description

This pull request introduced negative half-cell models to PyBaMM, by means of three major changes:
* The option `"working electrode": ["both", "negative", "positive"]` has been replaced with `"half-cell": ["false", "true]`. The positive electrode is always considered to be the working electrode, which has always been the case but is now made explicit.
* Two new parameter sets - `Ecker2015_graphite_halfcell` and `OKane2022_negative_halfcell` - have been added to work with the new option.
* The degradation mechanisms `SEI`, `SEI on cracks` and `lithium plating` can now be used on either electrode by specifying them as a 2-tuple, in the same way as is already done for some other options. Unlike for other options, however, passing only a single string instead of a tuple will put that behaviour on the negative electrode only. 

Fixes #2730 

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [x] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [x] No style issues: `$ pre-commit run` (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [x] All tests pass: `$ python run-tests.py --all`
- [x] The documentation builds: `$ python run-tests.py --doctest`

You can run unit and doctests together at once, using `$ python run-tests.py --quick`.

## Further checks:

- [x] Code is commented, particularly in hard-to-understand areas
- [x] Tests added that prove fix is effective or that feature works

DrSOKane commented:
Hi all, I'm sorry this is so slow! My local imageio isn't working for some reason and some of the tests depend on that, so I'm having to do some of the testing and debugging online instead of locally, which takes much longer.

Needless to say, this PR proposes a lot of big changes, but the one I would really appreciate external input on is the code block I added to base_battery_model.py to convert string inputs into tuples. Controversially, this alters extra_options and not default_options, which affects how the following code blocks - the ones that change default options - work.

brosaplanella commented on CHANGELOG.md:
```diff
@@ -7,11 +7,23 @@
 ## Breaking changes
 
 - Numpy functions now work with PyBaMM symbols (e.g. `np.exp(pybamm.Symbol("a"))` returns `pybamm.Exp(pybamm.Symbol("a"))`). This means that parameter functions can be specified using numpy functions instead of pybamm functions. Additionally, combining numpy arrays with pybamm objects now works (the numpy array is converted to a pybamm array) ([#3205](https://github.com/pybamm-team/PyBaMM/pull/32 ...
+- Option `working electrode` has been replaced with `half-cell`, which can only be `true` or `false`. The positive electrode is always considered to be the working electrode. For negative half-cells, lithium metal is now considered the negative electrode and the negative electrode material to be the positive electrode. ([#3198](https://github.com/pybamm-team/PyBaMM/pull/3198))
+- If `options["half-cell"] == "false"` and either `SEI`, `SEI on cracks` or `lithium plating` are not provided as tuples, they are automatically made into tuples. This directly modifies `extra_options`, not `default_options` to ensure the other changes `default_options` still happen when required. ([#3198](https://github.com/pybamm-team/PyBaMM/pull/3198))
+- Added option to use an empirical hysteresis model for the diffusivity and exchange-current density ([#3194](https://github.com/pybamm-team/PyBaMM/pull/3194))
+- PyBaMM now has optional dependencies that can be installed with the pattern `pip install pybamm[option]` e.g. `pybamm[plot]` ([#3044](https://github.com/pybamm-team/PyBaMM/pull/3044))
+- Double-layer capacity can now be provided as a function of temperature ([#3174](https://github.com/pybamm-team/PyBaMM/pull/3174))
+- `pybamm_install_jax` is deprecated. It is now replaced with `pip install pybamm[jax]` ([#3163](https://github.com/pybamm-team/PyBaMM/pull/3163))
+- PyBaMM now has optional dependencies that can be installed with the pattern `pip install pybamm[option]` e.g. `pybamm[plot]` ([#3044](https://github.com/pybamm-team/PyBaMM/pull/3044))
```
Some items are duplicated, probably due to merging other branches. Just check they only appear once and they are sorted in descending order of PR number.

brosaplanella commented on pybamm/input/parameters/lithium_ion/OKane2022_negative_halfcell.py:
Can you include the material rather than `negative` in the name?

brosaplanella commented on pybamm/models/full_battery_models/base_battery_model.py:
```diff
@@ -546,7 +563,7 @@ def __init__(self, extra_options):
         # Check options are valid
         for option, value in options.items():
-            if option in ["working electrode"]:
+            if option in ["half-cell"]:  # is this exception still necessary?
                 pass
```
Let's see what @tinosulzer says, as he implemented the original `working electrode` model. If you remove this line, do all tests still work?

brosaplanella commented on pybamm/models/full_battery_models/lithium_ion/base_lithium_ion_model.py:
```diff
@@ -159,13 +159,17 @@ def set_degradation_variables(self):
-        n_Li_lost_sei = self.variables["Loss of lithium to SEI [mol]"]
-        n_Li_lost_reactions = n_Li_lost_sei
-        if "negative electrode" in domains:
+        n_Li_lost_neg_sei = self.variables["Loss of lithium to negative SEI [mol]"]
+        n_Li_lost_pos_sei = self.variables["Loss of lithium to positive SEI [mol]"]
+        n_Li_lost_reactions = n_Li_lost_neg_sei + n_Li_lost_pos_sei
```
Could you include these in the loop below, or is there a reason to keep them separate?

DrSOKane replied:
It has to be separate because SEI can also grow on lithium metal electrodes, but that's not included in the loop because `domains` is defined on line 109 to be only whole cell domains.

brosaplanella commented on pybamm/models/submodels/interface/base_interface.py:
```diff
@@ -372,8 +372,9 @@ def _get_standard_sei_film_overpotential_variables(self, eta_sei):
         variables = {
-            f"{Phase_name}SEI film overpotential [V]": eta_sei,
-            f"X-averaged {phase_name}SEI film overpotential [V]": eta_sei_av,
+            f"{Domain} electrode {Phase_name}SEI film overpotential [V]": eta_sei,
```
Should this be `phase_name` (like the line below)?

brosaplanella commented on tests/unit/test_parameters/test_parameter_sets/test_OKane2022_negative_halfcell.py:
Do we also need one for Ecker's half cell parameters?

DrSOKane replied:
There isn't one for Ecker's full cell parameters so I didn't write one. I can write both if you wish

brosaplanella replied:
If you can that would be great, thanks!

brosaplanella reviewed (commented):
Great work, @DrSOKane! Just a few very minor comments

DrSOKane commented:
The tests finally pass!! @tinosulzer @rtimms

brosaplanella reviewed (approved):
Looks great, thanks Simon!

tinosulzer reviewed (commented):
Looks good, thanks @DrSOKane ! 

A couple of comments:
- Is a helpful error raised if the user tries to do `{"working electrode": "positive"}`?
- It looks like the SEI variables now require a domain ("Negative electrode SEI thickness [m]" instead of "SEI thickness [m]"). Does that mean previous scripts with "SEI thickness [m]" will break?

rtimms commented:
thanks @DrSOKane ! would be great to get this merged before the next release if you have time to respond to valentin's comments above

tinosulzer reviewed (changes requested):
Upon further considering I think we should keep "working electrode": "positive" to avoid unnecessarily breaking people's code, and just error with "working electrode": "negative" with an error message that explains you always have to use "positive" and change the parameters. In future if we have a symmetric cell that can be "working electrode": "none"

DrSOKane commented:
Hi @tinosulzer and @rtimms, I'm back from holiday and made this a priority.

All variables pertaining to the `SEI`, `SEI on cracks` and `lithium plating` submodels do indeed have domains now, and I've added this to the list of breaking changes.

Having thought about it, I agree that changing back to the `working electrode` syntax is the best way to go, so I've gone through every instance and changed it back. An `OptionError` is now raised if you attempt to set `working electrode` to be `negative`.

Finally, I've added an example notebook to show how to use the half-cell models.

brosaplanella commented:
Happy to merge once the notebook conflicts are resolved

brosaplanella commented:
@agriyakhetarpal do you know why the doc test is failing?

agriyakhetarpal commented:
Yes, I think this is a bug that I have noted earlier (#3389). We should mark it at a higher priority now I guess

DrSOKane commented:
> Yes, I think this is a bug that I have noted earlier (#3389). We should mark it at a higher priority now I guess
> 
> Edit: sorry, I read the logs too quickly. It looks like a different warning, I will debug and leave a review below.

The underline in parameter_sets.rst is to short to accommodate the name of the new parameter set, so I extended the underline.

brosaplanella commented:
Fantastic work @DrSOKane!
